### DISCUSSION TITLE
Using custom build task in net5.0 project and building the project through VS

### DISCUSSION THREAD

**Discussion by persiandeveloper** [dotnet/msbuild#6692] [Q&A]
Recently I had issues using a custom build task in a .net core 5.0 project. I got an answer from [this ](https://docs.microsoft.com/en-us/answers/questions/459914/custom-buld-task-when-used-in-csproject-throws-39c.html?childToView=470025#answer-470025) . 

My Code is accessible [here ](https://github.com/persiandeveloper/custom-build-task). 

If I clean the target project and try to rebuild,  I have to clear the following part first : 

`<Target Name="PreBuild" BeforeTargets="PreBuildEvent">
    <SnykTaskFile Location="$(ProjectDir)" />
  </Target>`

Then I have to build it, otherwise I get the following error : 

`The "SnykTaskFile" task could not be loaded from the assembly D:\2021\customTask\Application\bin\Debug\net5.0\SnykTask.dll`

Which makes sense because the 'SnykTask.dll' is not there. 

My question is, how should I reference the CustomTask library into target project so I do not need to the the above steps each time. 

But there is more serious issue when I build this same solution in another pc, I get the following error : 

```
The "SnykTaskFile" task failed unexpectedly.
System.IO.FileNotFoundException : Could not load file or assembly 'System.Runtime, Version = 4.2.1....
```

In this case, I have to clean project, remove **`<Target Name="PreBuild" BeforeTargets="PreBuildEvent">`**, and build it, build fails in Visual Studio and still shows the above message, how ever "dotnet build" command works properly. 

**Update**

I fund out that the issue is "System.Text.Json" If use "newtonsoft" it works. However still do not know how can I use "System.Text.Json" not "newtonsoft"

**Comment by KalleOlaviNiemitalo** [comment]
Perhaps you could move the JSON code to a separate executable program and start that from SnykTask. This way, SnykTask could target .NET Standard 2.0, and the separate process would be able to use .NET 5.0 even if MSBuild is running on .NET Framework. This might require large changes in your code, though.

I suggest adding `ReferenceOutputAssembly="false"` metadata to the ProjectReference so that the MSBuild will build SnykTask but not let the C# compiler reference SnykTask.dll when compiling Application.
